src/plugin/csi/node.go

```diff
@@ -0,0 +1,239 @@
+	var fi os.FileInfo
+	for {
+		var err error
+		fi, err = os.Stat(path)
+		if err == nil {
+			// TODO(msteffen): handle case where data is partially-synced.
+			if err := os.RemoveAll(path); err != nil {
+				return nil, status.Error(codes.Internal,
+					fmt.Sprintf("could not remove existing dir at %q: %v", path, err.Error()))
+			}
+			continue // retry stat()
+		} else if !os.IsNotExist(err) {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+	if err := os.MkdirAll(path, 0750); err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	if !fi.IsDir() {
+		return nil, status.Error(codes.Internal, fmt.Sprintf(
+			"cannot stage volume at %q (%s): it is not a directory",
+			path, fi.Mode()))
+	}
+
+	// sync pachyderm data to staging volume
+	p := filesync.NewPuller()
+	if err := p.Pull(c, path, repo, commit, commitPath,
+		false /*lazy*/, false, /*empty*/
+		concurrency,
+		nil /*statsTree*/, "" /*statsRoot*/); err != nil {
+		return nil, status.Error(codes.Internal,
+			fmt.Sprintf("could not download Pachyderm data at %s@%s:%s: %v", repo, commit, path, err))
+	}
+	return &csi.NodeStageVolumeResponse{}, nil
+}
+
+// NodePublishVolume implements the corresponding endpoint in the CSI service
+// definition. It's used to bind mount the global directory on a container
+// directory.
+func (s *NodeSvc) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolumeRequest) (*csi.NodePublishVolumeResponse, error) {
+	// Check arguments
+	if req.GetVolumeCapability() == nil {
+		return nil, status.Error(codes.InvalidArgument, "Volume capability missing in request")
+	}
+	if len(req.GetVolumeId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume ID missing in request")
+	}
+	if len(req.GetTargetPath()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Target path missing in request")
+	}
+
+	// Create mount point (or re-use existing mount point if it's there)
+	fi, err := os.Stat(req.TargetPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if err = os.MkdirAll(req.TargetPath, 0750); err != nil {
+				return nil, status.Error(codes.Internal, err.Error())
+			}
+		} else {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+	if !fi.IsDir() {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("%q (%s) is not a directory", req.TargetPath, fi.Mode()))
+	}
+
+	// log mount information
+	log.Printf("target %v\nfstype %v\nreadonly %v\nvolumeId %v\nattributes %v\nmountflags %v\n",
+		req.TargetPath, req.VolumeCapability.GetMount().FsType, req.Readonly,
+		req.VolumeId, req.VolumeContext, req.VolumeCapability.GetMount().MountFlags)
+
+	// finally, mount 'path' to 'targetPath', exposing pach data to pod
+	path := pathlib.Join(provisionRoot + req.VolumeId)
+	var mntOpts uintptr = syscall.MS_BIND
+	if req.Readonly {
+		mntOpts |= syscall.MS_RDONLY
+	}
+	if err := syscall.Mount(path, req.TargetPath, "", mntOpts, ""); err != nil {
+		return nil, err
+	}
```

Comment:
We had issues in the past with needing to be privileged to perform a bind-mount. Are these containers already privileged so that's not a concern? Also maybe there's no other way to do it.